File: example/1d-linear-convection/weno3/julia/03c/src/time_integration/rk3.jl
# src/time_integration/rk3.jl

mutable struct RK3Integrator
    base::TimeIntegratorBase
end

function RK3Integrator(cfd::Any)
    RK3Integrator(TimeIntegratorBase(cfd))
end

function step(integrator::RK3Integrator, dt::Float64)
    base = integrator.base
    # Stage 1
    compute_residual(base)
    u1 = copy(base.solution.u)
    for i in base.domain.ist:(base.domain.ied - 1)
        j = map_idx(base, i)
        u1[i] += dt * base.solution.res[j]
    end
    base.solution.u .= u1
    apply_boundary(base)
    # Stage 2
    compute_residual(base)
    u2 = copy(base.solution.u)
    for i in base.domain.ist:(base.domain.ied - 1)
        j = map_idx(base, i)
        u2[i] = 0.75 * base.solution.un[i] +
                0.25 * base.solution.u[i] +
                0.25 * dt * base.solution.res[j]
    end
    base.solution.u .= u2
    apply_boundary(base)
    # Stage 3
    compute_residual(base)
    c1, c2, c3 = 1.0/3.0, 2.0/3.0, 2.0/3.0
    for i in base.domain.ist:(base.domain.ied - 1)
        j = map_idx(base, i)
        base.solution.u[i] = c1 * base.solution.un[i] +
                             c2 * base.solution.u[i] +
                             c3 * dt * base.solution.res[j]
    end
    apply_boundary(base)
    update_old_field(base.solution)
end

File: example/1d-linear-convection/weno3/python-v1/01c/time_integration/rk3.py
# time_integration/rk3.py

from .base import TimeIntegrator
from cfd_registry import register_component

@register_component('integrator', 'rk3')
class RK3Integrator(TimeIntegrator):
    """3阶SSPRK3（强稳定保号RK3）"""
    def step(self, dt):
        # Stage 1
        self.compute_residual()
        u1 = self.solution.u.copy()
        for i in range(self.domain.ist, self.domain.ied):
            j = self.map_idx(i)
            u1[i] += dt * self.solution.res[j]
        self.solution.u[:] = u1
        self.apply_boundary()

        # Stage 2
        self.compute_residual()
        u2 = self.solution.u.copy()
        for i in range(self.domain.ist, self.domain.ied):
            j = self.map_idx(i)
            u2[i] = (
                0.75 * self.solution.un[i] +
                0.25 * self.solution.u[i] +
                0.25 * dt * self.solution.res[j]
            )
        self.solution.u[:] = u2
        self.apply_boundary()

        # Stage 3
        self.compute_residual()
        c1, c2, c3 = 1.0/3.0, 2.0/3.0, 2.0/3.0
        for i in range(self.domain.ist, self.domain.ied):
            j = self.map_idx(i)
            self.solution.u[i] = (
                c1 * self.solution.un[i] +
                c2 * self.solution.u[i] +
                c3 * dt * self.solution.res[j]
            )
        self.apply_boundary()
        self.solution.update_old_field()